### PR TITLE
automatic types suggestion, closes #2

### LINKED ISSUE #2: Add method to guess the scitypes of a table
Would output a dictionary `changes` on column names, whose values are the suggested scitype for the column with that name. The user could could then edit the dictionary and then effect the type coercions with `coerce(changes, data)`. 

So, if the new method is called `suggest_scitypes` (better name?) the workflow would be something like:

```julia
juliia> X = (book=["red", "white", "blue", "blue"], 
               number=[0, 1, 1, 0, 1],
               is_broken=[0, 1, 1, 0])

julia> changes = suggest_scitypes(X)
Dict{Symbol,Any} with 3 entries:
  :book => Multiclass
  :number => OrderedFactor
  :is_broken => OrderedFactor

julia> changes[:number] = :Count 
julia> X_fixed = coerce(changes, X)
```

### PR DESCRIPTION
**Pull request by tlienart**

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -6,6 +6,7 @@ version = "0.2.0"
 [deps]
 InteractiveUtils = "b77e0a4c-d291-57a0-90e8-8db25a27a240"
 Requires = "ae029012-a4dd-5104-9daa-d747884805df"
+Tables = "bd369af6-aec1-5ad0-b16a-f7cc5008161c"
```
Comment (ablaom) [inline Project.toml]: Currently Tables.jl is an optional dependency. I can see an argument for making it a real dependency, as it is quite light-weight. Post an issue that the docs will need to be updated.

File: src/conventions/mlj/auto_types.jl [hunk 2]
```diff
@@ -0,0 +1,51 @@
+export auto_types
+
+"""
+suggest_scitype(type, col, nrows)
+
+For a column `col` with element type `type` and `nrows` rows, try to suggest an appropriate
+scitype in cases which would otherwise be typed as "unknown" namely when there are few
```
Comment (ablaom) [inline src/conventions/mlj/auto_types.jl]: "unknown" -> `Unkown`

File: src/conventions/mlj/auto_types.jl [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+function suggest_scitype(type, col, nrows)
+   # TODO: deal with missing/union (?)
+   unique_vals  = unique(skipmissing(col))
+   nunique_vals = length(unique_vals)
+   # Heuristic 1
+   if nunique_vals ≤ 3 && nrows ≥ 5
+      return Multiclass{nunique_vals}
+   # Heuristic 2
+   elseif nunique_vals ≤ max(min(0.1*nrows, 100), 4)
+      type <: Real && return OrderedFactor{nunique_vals}
```
Comment (ablaom) [inline src/conventions/mlj/auto_types.jl]: Presently, coerce to `OrderedFactor` will fail on `Real` types, which will need to be changed.

File: src/conventions/mlj/auto_types.jl [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+Otherwise fall back to the default typing.
+"""
+function suggest_scitype(type, col, nrows)
+   # TODO: deal with missing/union (?)
+   unique_vals  = unique(skipmissing(col))
+   nunique_vals = length(unique_vals)
+   # Heuristic 1
+   if nunique_vals ≤ 3 && nrows ≥ 5
+      return Multiclass{nunique_vals}
```
Comment (ablaom) [inline src/conventions/mlj/auto_types.jl]: When  you call `coerce(v, Multiclass)` there is no need to specify the number of levels, and the method will probably fail if you do. So also probably not appropriate to return the parameter here and below.

**Comment by tlienart** [comment]
Hmm I should probably have it so that if `auto_types` detects missing value it suggests a `Union{Missing, OrderedFactor}` or similar, will update the PR soon for this.

**Comment by ablaom** [comment]
Well, not necessarily. Currently, coerce([1,2, missing], OrderedFactor) works fine. It just issues a warning to say that the new scitype is `Union{Missing,OrderedFactor{2}}`. 

We already have a method for inspecting scitypes of a column or table. I see the raison d'etre for `auto_types` is for giving us a (editable) dictionary to pass to `coerce`, rather than reporting stuff about levels and missing values. 

No?

**Comment by tlienart** [comment]
I didn't get your last comment about adding `istable`  check in `conventions/mlj/mlj.jl` it didn't seem necessary?

I did however enact a few changes in there, hope you won't mind, the way things are written now should be a bit more robust and more informative, in particular I added a warning if the user coerces a Char to either float or int as that could lead to unexpected results.

**Comment by ablaom** [comment]
Thanks indeed for these improvements. 

- Perhaps one sanity check would be to look at suggestion for the MLJ built-in Ames data set, which has 76 features of mixed type. This is for our own info, not a formal test:

```julia
julia> using ScientificTypes

julia> using CSV; X = load_ames();

julia> auto_types(X)
Dict{Symbol,Type{#s2} where #s2<:Union{Missing, Found}} with 76 entries:
  :YearRemodAdd  => OrderedFactor
  :Street        => Multiclass
  :PavedDrive    => Multiclass
  :x2ndFlrSF     => Count
  :TotalBsmtSF   => Continuous
  :KitchenQual   => Multiclass
  :LotFrontage   => Continuous
  :YrSold        => OrderedFactor
  :GarageYrBlt   => OrderedFactor
  :GarageCond    => Multiclass
  :target        => Continuous
  :SaleCondition => Multiclass
  :GarageFinish  => Multiclass
  :BsmtExposure  => Multiclass
  :RoofMatl      => Multiclass
  :HalfBath      => Multiclass
  :MSZoning      => Multiclass
  :MoSold        => OrderedFactor
  :ExterCond     => Multiclass
  :ExterQual     => Multiclass
  :GarageType    => Multiclass
  :BsmtFinType1  => Multiclass
  ⋮              => ⋮
```
For instance, I was a bit surprised at the suggestion `:x2ndFlrSF     => Count` which is square footage. Of course no heuristic is foolproof and maybe this is just fine. (Later we can *train* this suggester on actual cases; I think there is already work along these lines at Turing ) I am attaching a summary of a description of the data.

- Maybe, as a kwarg option, report only columns for which the suggested coercion is actually needed (by comparing with `scitype(existing_colummn)`. What do you think?

- Maybe "autotype" is readable enough without the underscore. We already have `scitype`, `eltype`, and so forth. You make an executive decision. 

- We'll need the REAME updated, which is generated from docs/src/index.jl (a little messy at present - you run make.jl and export the markdown from the generated notebook - I can do this if you add the text.)

**Comment by tlienart** [comment]
## AMES

* `:x2ndFlrSF => Count` this is not due to the Heuristic  (no count of unique values here) but just to the mlj convention `Integer => Count`, I did not touch this; are you ssuggesting it should be changed / there should be more logic here?

Where it does do something is for instance

* `:Street => Multiclass` this is fine as there are only two values here, `Pave` and `Grvl`
* `:OverallCond => OrderedFactor` this is as expected too as the values are `1...9`

I would personally be against much more logic for type guesses **or** we could make any more sophisticated guessing toggle-able and then the user could select a bunch of standard additional rules they'd want to use. I'm not sure what these rules could be but for instance:

* Consider all `Integer` with many values as `Continuous` instead of `Count`
* Consider all `String` types as Multiclass no matter how many there are
* ...

But such choices would only make senses in specific scenarios.

Thus far we've only used a heuristic along the lines of "_if there's relatively few unique values, use a finite type_" which is easy to explain, seems fair, and should work in a large number of cases.

## additional comments

ok for all, WIP

* [ ] add kwarg to only report the types that would have been guessed
* [ ] `auto_type` to `autotype` ok
* [ ] readme update

**Comment by tlienart** [comment]
As far as I'm concerned this is ready to be merged into `dev` if you're ok with comments above.

**Comment by ablaom** [comment]
Many thanks for the changes and docs!

After reading your documentation I am left a little confused. For instance, you say "in cases which would otherwise be typed as `Unknown`" and yet autotypes does already consider new types for `Real` data (which includes `AbstractFloat` which has scitype `Continuous`, not `Unknown`). Not that this is a bad thing, just that it's not clear.

I am also confused by " The *mlj* convention comes with a `autotype` function which adds to
the above rules ...". I guess we differ in how we view the role of  `autotypes`. Correct me if I am mistaken, but you regard it as supplementing the mlj convention, whereas I would say it should be independent of conventions altogether. Here is a vector. What would you guess is the scientific type (ie, most likely interpretation)? There may be overlap with the *mlj* convention, but this convention is necessarily imperfect as it is constrained by the rule "each julia object has a unique scientific type".

**Comment by tlienart** [comment]
Ok I'm left a bit confused here, apologies if I'm being dense.

-  doc clarity, ok

With respect to where `autotype` stands with regards to the mlj convention I think it's important we figure that out precisely. I agree that `autotype` could just be a function that lives on its own, independent of a convention but that requires passing it a number of _rules_:

- rules for when there are few values
- generic rules such as Integer -> continuous
- default rules

At the moment only the first one is suggested; for the last two the fallback is the MLJ convention (as far as I'm aware).

So I have two suggestions:

### Suggestion 1

(I may prefer this one but it requires significant changes)

* Make conventions into **const dictionaries** of `Type => Scitype` which will be used as fallback (for instance, currently, the `Integer=>Count` in `autotype` is an implicit use of this fallback)
* Allow `autotype` to live on its own (separate of any convention), with
   * a kwarg that picks a convention (fallback) e.g. MLJ
   * other kwargs for pre-implemented generic rules such as Integer->Continuous or Few values  -> Finite
   *  a way too pass user-defined rules (e.g. if they want to specify their own "few values" -> "Finite"

### Suggestion 2

* Keep things as they are
* A "new convention" would implement its own `autotype`
* Add keyword argument to `autotype` for aditionnal generic rules as above

Could you let me know which way you'd prefer to go? If the former, I think I may branch off this PR and open an other PR.

**Comment by ablaom** [comment]
I expect I am probably missing something here because I'm not the one actuallly writing the implementation :-) Let's chat about this again at our next meeting. Below are some thoughts. 


> Make conventions into const ordered-dictionaries of Type => Scitype which will be used as fallback (for instance, currently, the Integer=>Count in autotype is an implicit use of this fallback)

This would be nice, but it doesn't work. The scitype of an object cannot be determined from the type of the object. For example, to determine if a `CategoricalValue` is ordered or not, I must look at the `ordered` attribute. The type does not tell me. While personally I think this is a design flaw (see https://github.com/JuliaData/CategoricalArrays.jl/issues/184) there's no guarantee this problem does not reappear as we expand the convention. 

Ultimately, I view the mlj convention as an implementation detail that, with the help `schema`, `coerce` and `autotypes`, the casual user can ignore. I don't see that `autotypes` as a convention in this sense. It is an intelligent system that tries to guess the intended scientific interpretation of some column from a combination of the machine type(s) used and the actual values. (By contrast, the `scitype` of a column is just `AbstractVector{T}` where `T` is the union of the element scitypes). We could make this system highly customisable with user-specified rules, but I would say this out of scope for now. I suggest we just implement some quick and dirty method that saves the user having to construct his `coerce` dictionary completely from scratch, which is what you have done already. I just thought we could improve it by adding some sensible rules for Integers.

> Suggestion 2

> Keep things as they are
> A "new convention" would implement its own autotype
> Add keyword argument to autotype for aditionnal generic rules as above

I don't understand why `autotype` needs to be implemented for each convention in any scenario. Actually, we should move the existing code to `tables.jl` with a hook for the method in `ScientficTypes.jl` like we have for `schema`. It is true that the *output* of `autotypes` depends on the active convention, in the special case that `suggestions_only` is true (maybe `changes_only` is better??) but I don't see why the method itself should care.

**Comment by tlienart** [comment]
Ok I believe I understand now, thank you. I'll push changes before the call so that we can discuss them live if necessary.

**Comment by tlienart** [comment]
Ok I think this is better now (I hope!)

1. it's not part of the convention anymore
2. a set of rules can be specified, the base one being `few_to_finite`; they're applied one after the other so the precedence matters but I think that makes sense
3. if no rule applies, the scitype is returned (this amounts to the convention fallback which, usually, will be mlj)

The tests (for autotype) are clearer and better I think and should showcase how this works. 

I haven't changed the docs in `index.jl` but will do so once we're ok with the changes.

**Comment by ablaom** [comment]
Thanks.

The new method inspects the *scitype* of the column, to be used as a criterion for suggesting the scitype for coercion. What I had in mind is a method that bases its suggestion purely on the machine type. So,  eg, `<:Integer` goes to `Countinuous` if .... but  `OrderedFactor` if ...... That way the operative scitype convention is immaterial (except when deciding if this is a change).

**Comment by tlienart** [comment]
Either I don't understand this or I don't like it. Effectively, doing what I think you're asking would amount to adding a bunch of default rules such as:

* `float_to_continuous`
* `int_to_continuous`
* `grayarray_to_grayimage`
* `colorarray_to_colorimage`

which is exactly what a convention such as MLJ already defines. So I guess we could bypass the convention completely and effectively hardcode the MLJ convention again in autotype but I don't see the point. 

Could you please clarify why you don't want this, possibly with an example of a scenario I'm missing, and how you envisage it being done?

**Comment by ablaom** [comment]
Okay, after our call and reflection, I'm happy to go with this as is, if you are. Update the doc/docstring and merge - many thanks.

**Pull request merged**